### DISCUSSION TITLE
Ensure developers can modify and extend Wagtail Admin UI using only Django/Python code

### DISCUSSION THREAD

**Discussion by brylie** [wagtail/wagtail#8128] [Ideas]
As Wagtail moves forward with improvements to the Admin user interface, the project invests in an increasing amount of JavaScript. The driving idea is that Wagtail Admin UI should feel "[less like a webform and more like an app](https://www.youtube.com/watch?v=CQ7Gx8b7ac4&t=381s)", as has been articulated on the Wagtail draft 2023 roadmap. While this goal is undoubtedly laudable, we should clearly commit to maintaining the simplicity that Django and Wagtail developers have come to appreciate with the frameworks.

Thankfully, the Wagtail 2023 (draft) roadmap also includes a pillar of "[sophistication without compromising simplicity](https://www.youtube.com/watch?v=CQ7Gx8b7ac4&t=424s)."

## Proposal

We should make a _clear and open commitment to maintaining developer simplicity_ while expanding Wagtail's dynamic capabilities. This principle has already been justified for [preserving backward compatibility and ensuring that Wagtail is a powerful CMS for Django/Python developers.](https://www.youtube.com/watch?v=nBNb7k0BMP4&t=2574s) while informing software architecture decisions. A good example comes from the [Telepath initiative](https://wagtail.io/blog/telepath/), which enhances the Wagtail core capabilities while [remaining transparent to the Wagtail developer](https://youtu.be/j9XDr9xtfPI?t=905).

Wagtail developers (who build downstream projects) should maintain the ability to modify and extend the Wagtail Admin UI and framework using only Django and Python code. This would imply:

- no JS build tools are required for downstream Wagtail development,
- code can be JS framework agnostic, 
- developers can choose whichever tools/libraries are suitable for their project

## Summary

In various contexts, the Wagtail Core development team and Torchbox have articulated a commitment for Wagtail to be friendly and familiar for Django/Python developers. It might just prove useful to have a clearly articulated version of this commitment in some fixed form (such as a "Wagtail design principles" document) to serve as a guideline for design and architecture decisions.

**Comment by gasman** [comment]
Thanks for the suggestion @brylie! Ever since putting together the "Zen of Wagtail" page for the docs, it's been at the back of my mind to do a 'design principles' document in the same vein - perhaps in the Contributing section of the docs - for guidelines specific to the development of Wagtail itself, rather than ones aimed at site implementers. Will see what I can come up with...

A couple more that spring to mind immediately:

* extend Django, don't replace it
* hide features that add complexity (e.g. workflows, collections), until the user is actively making use of them
* don't rely on config flags
* encourage best practices out of the box (e.g. not opening links in new windows - if developers want to overrule that, it's okay to make them jump through hoops)

**Comment by thibaudcolas** [comment]
I’ll convert this to GitHub discussions now as I believe we could use more people talking about this, and there isn’t really any one resolution to this. We might end up having a document that clarifies our vision in that respect, or we might just keep it as an informal design principle that you have to learn on the job while working on UI customisations. Even if we did make a document that could "resolve" this issue, I suspect we’d rather keep the discussion going for people to feed back and improve said document over time.

We’ve also been discussing this in #7689, and #7739 (in particular see the results of [this poll](https://github.com/wagtail/wagtail/discussions/7739#discussioncomment-2244359)).

**Comment by lb-** [comment]
This is a great summary of what I have been thinking about for Wagtail, I have kind of summed it up as Wagtail should aim to still be 'just' Django.

When it comes to frontend end interactivity, I think we should also aim for Wagtail to work well with existing Django abstractions. HTML templates (overriding, extending, including) should work, using widget `atrrs` should work. Also leveraging JS DOM events better will mean that other JS libraries can easily interact with each other irrespective of their tooling. 

Previous comment about DOM Events. https://github.com/wagtail/wagtail/discussions/7689#discussioncomment-1666918

Related investigation into a lightweight framework https://github.com/wagtail/wagtail/discussions/7689#discussioncomment-2037913
